Lplb/session management (#382)
Unity-Technologies/com.unity.multiplayer.docs: docs/advanced-topics/session-management.md

---
id: session-management
title: Session Management
---
In a multiplayer game, we might want to have some session management to allow players to correctly reconnect to the game. What this exactly means might vary depending on the game, but the core idea is that there is some data that needs to be kept when a player disconnects, and assigned back to them when they reconnect. We must also make sure that the data is reassigned to the right player.

## Linking data to players

> To reassign data to the right player when they reconnect, we need a way to link this data to the player. The clientId generated by NGO cannot be used, however, since it is generated when a player connects and is disposed of when they disconnect. It can then be assigned to a different player when if they connect to the server after the first one disconnected. To properly link data to a specific player, we need an ID that is not tied to the current session, that persists through time, and that is unique to that player. A few solutions can be used, such as a login system with unique player accounts, or simply a Globally Unique Identifier (GUID). A GUID can be generated via `System.Guid.NewGuid()` and then saved to the `PlayerPrefs`, for example.

Reviewer: ```suggestion
To reassign data to the right player when they reconnect, we need a way to link this data to the player. The clientId generated by NGO cannot be used, however, since it is generated when a player connects and is disposed of when they disconnect. It can then be assigned to a different player when if they connect to the server after the first one disconnected. To properly link data to a specific player, we need an ID that is not tied to the current connection, that persists through time, and that is unique to that player. A few solutions can be used, such as a login system with unique player accounts, or simply a Globally Unique Identifier (GUID). A GUID can be generated via `System.Guid.NewGuid()` and then saved to the `PlayerPrefs` client side, for example.
```

When we have such a way to uniquely identify a player, we can then use it to map each player's data that would be needed when reconnecting, such as the current state of their character for example. Then, we need to make sure this data is up to date and to keep it after a player disconnects. We could add a timeout after which we can purge the data or we could decide to clear all data only when a session completes. And finally, once we have up to date data that we can link to a specific player who disconnected earlier, we can now handle reconnection properly.

## Handling reconnection

> In cases where we use a Player Object, a new Default Player Prefab will be automatically spawned when any player connects to the game (including if they are reconnecting). We can then use the player's data to update that object so that its state returns to what it was before disconnecting. In those cases, we would need to keep all the important data that we want to restore and map it to the player using our identification system.

Reviewer: ```suggestion
In cases where we use a Player Object, a new Default Player Prefab will be automatically spawned when any player connects to the game (including if they are reconnecting). We can then use the player's previously saved session data to update that object so that its state returns to what it was before disconnecting. In those cases, we would need to keep all the important data that we want to restore and map it to the player using our identification system.
```

> In cases where we use the ownership model, we could instead keep the objects that a player owns when they disconnects, and set the reconnected player as their new owner. To accomplish this, the only data we would need to keep would be the mapping between those objects and their owning player's identifier, then when a player reconnects we could use this mapping to set them as the new owner.

Reviewer: "ownership model" ? Do we mean Netcode's ownership?

Author: Yes, I used the term defined [here](https://docs-multiplayer.unity3d.com/docs/develop/basics/networkobject). Maybe I should add a link to that page at the beginning of this section?

Author: In that case, for consistency we should probably rewrite this sentence from that other page: "If you want a client to control multiple objects use the ownership model instead."

Reviewer: why do you call this a "model"? To me it's more of an API or a feature. Prediction is a lag compensation "model", there's multiple ways to implement it, but the principle is the same. To me, a model is more like a "template" or a pattern, more than an SDK feature.

Author: What I mean is that the term "model" to distinguish between using a PlayerObject vs using manually specified Ownership for NetworkObjects comes from [here](https://docs-multiplayer.unity3d.com/docs/develop/basics/networkobject/index.html). I agree with what you are saying, that another word might be less confusing, but we should also change it in that other page if we want to be consistant

Reviewer: @s-omeilia-unity ^^^

Author: Would "approach" be a better term for that distinction?

## Example using the Player Object model

> Here is an example from the Boss Room sample, showing some simple session management. The game uses the Player Object model and a GUID to identify unique players. In that example, each player's data is saved in the `m_ClientData` dictionary, and each of their clientId is mapped to their Guid in `m_ClientIDToGuid`. The saved data is represented by the generic type `T`, which is a struct implementing the `ISessionPlayerData` interface. This struct contains all the data required. An implementation could contain, for example, the character's current hit points, position, class, etc. The interface has two properties, `IsConnected` and `ClientID`, and the method `Reinitialize`. `IsConnected` is used in the `OnClientApprovalCheck` method to differentiate between a duplicate connection and a reconnection, and is set when a player connects to the game and is unset when they disconnect. `ClientID` is used to link the player's session data with their current ClientID. It is updated when they connect or reconnect. Finally, `Reinitialize` is called in the `OnSessionEnded` method for each player's session data. This should reset the data of the player's session to represent a new session.

Reviewer: Do we really need to describe the code we wrote? The code is pretty self describing.

Reviewer: The "Why" is usually the most important part

Reviewer 2: is the code self describing to everyone or just if you know what it does...is this something that is useful if it is then let's leave it in place

Reviewer: This page is for programmers, if they can read code, they might not need chunks of this doc (what do you think @LPLafontaineB ?)
It might be worth it to talk about how boss room uses it, more than what the session manager itself does (as it's designed to be reused)

Author: Rereading it, I think there is definitely part of it that's not useful and could be cut, while adding something explaining the reasons for our choices and how it is used inside BossRoom a bit better

Reviewer 2: I am going to ask the question again, does leaving the extra in hinder an experienced coder as much as it may help someone learning? that's the question. Remember that while the target audience may be experienced its not a bad thing if we also help the others.

Author: In this case, it's more that I was repeating myself. This extra info was describing something that is already explained by the comments in the code itself, or that is self-describing code. I don't think removing those parts would hinder someone learning.

Author: Although I did have my nose inside this system for weeks, so there might be some parts of it that seem evident to me but are not

Reviewer: A bigger conversation could be about how to address experienced coders vs beginners. Could we have a "beginner explanation" collapsible section? Something that would allow experienced devs to skip to the part they want?
Or have a TLDR section at the top with the essentials for experienced devs and then the rest of the articles for beginners? It's always an issue I have with some of these docs, is being able to get to the information I really need without getting the whole world explained to me yet again.

Author: For this specific PR however, should we wait for this conversation's conclusion? In this case I feel that the extra information I removed is mostly clutter and doesn't really help to explain the topic, so either way I think it should be fine on that front, but I'm open to add more details if we prefer going that way instead.

Reviewer: Very old article that kept popping up in my last 7 years of career creating and documenting code:
https://blog.codinghorror.com/code-tells-you-how-comments-tell-you-why/
Code tells you how, comments tell you why. This doc should be telling you the why and point to the code for the how. You don't want to reimplement our code in plain text, that's useless, gets out of date very quickly and our users don't need this (even non-programmers, as if they don't understand code, they won't understand code-in-plain-text). It's just adding easily outdated noise. Doc should be as straightforward as possible, else readers will just glance over it and miss the important bits.
If you think this adds to your code, if there's a trick you used that could be worth explaining, then sure (for example anything related to managing GUIDs and the discussions we've had in the auth channel). But implementation details should be skipped. Users could want to implement this their own way as well, they don't HAVE to implement it the way you did.

Author: Makes sense. In that case, I think we should keep it how it currently is

---
id: session-management
title: Session Management
---

You can use session management to ensure that when a player disconnects, some data is kept and accurately assigned back to that player when they reconnect.

## Linking data to players

To reassign data to the right player when they reconnect, we need to link this data to the player.

The **clientId** generated by Netcode for GameObjects (Netcode) cannot be used, because it generates when a player connects and is disposed of when they disconnect. That ID may then be assigned to a different player if they connect to the server after the first one disconnected.

To properly link data to a specific player, we need an ID that is not tied to the current connection, but persists through time and is unique to that player. Some options include
* A login system with unique player accounts
* A Globally Unique Identifier (GUID). For example, a GUID generated via `System.Guid.NewGuid()` and then saved to the `PlayerPrefs` client side.

> With this unique identifier, we can map each player's data that is needed when reconnecting, such as the current state of their character. We then ensure this data is up to date and kept after a player disconnects to properly reconnect the player.

Reviewer: ```suggestion
With this unique identifier, we can map each player's data that is needed when reconnecting, such as the current state of their character (last known position, health, etc). We then ensure this data is up to date and kept host side after a player disconnects to properly repopulate that player's data when they reconnect.
```

You can also decide to clear all data when a session completes or add a timeout to purge the data after a specific amount of time.

## Reconnection

> Reconnecting players depends on your game. For example, if we use a [Player Object](../basics/networkobject.md#player-objects), a new `Default Player Prefab`  automatically spawns when any player connects to the game (including when they reconnect). You can then use the player's previously saved session data to update that object so that its state returns to what it was before disconnecting. In those cases, we would need to keep all the important data that we want to restore and map it to the player using our identification system.

Reviewer 3: @LPLafontaineB Is there a way/suggestion/best practice to explain how to use the saved session data to update the object?

Author: I think so, yes. For example, in Boss Room, it is done in the OnNetworkSpawn() method of the object. The object then simply gets the data and applies it to itself. I'll find a way to phrase it and update this.

> In cases where we use don't use the Player Object approach and instead manually attribute client ownership to NetworkObjects, we could keep the objects that a player owns when they disconnect, and set the reconnected player as their new owner. To accomplish this, the only data we would need to keep would be the mapping between those objects and their owning player's identifier, then when a player reconnects we could use this mapping to set them as the new owner.

Reviewer 3: Similar to the above comment, do we have a suggestion/best practice on how to set up this mapping and setting them as the new owner?

> In cases where we use don't use the Player Object approach and instead manually attribute client ownership to `NetworkObject`(s), we could keep the objects that a player owns when they disconnect, and set the reconnected player as their new owner. To accomplish this, the only data we would need to keep would be the mapping between those objects and their owning player's identifier, then when a player reconnects we could use this mapping to set them as the new owner. This mapping can be as simple as a dictionary mapping the player identifier with the `NetworkObjectId`(s) of the `NetworkObject`(s) they own. Then, in the `OnClientConnectedCallback` from the `NetworkManager`, the server could set the ownership of these objects.

Reviewer: ```suggestion
In cases where we don't use the Player Object approach and instead manually attribute client ownership to `NetworkObject`(s), we could keep the objects that a player owns when they disconnect, and set the reconnected player as their new owner. To accomplish this, the only data we would need to keep would be the mapping between those objects and their owning player's identifier, then when a player reconnects we could use this mapping to set them as the new owner. This mapping can be as simple as a dictionary mapping the player identifier with the `NetworkObjectId`(s) of the `NetworkObject`(s) they own. Then, in the `OnClientConnectedCallback` from the `NetworkManager`, the server could set the ownership of these objects.
```

## Example using the Player Object approach
Here is an example from the Boss Room sample, showing some simple session management. The game uses the Player Object approach and a GUID to identify unique players.

```csharp reference
https://github.com/Unity-Technologies/com.unity.multiplayer.samples.coop/blob/develop/Packages/com.unity.multiplayer.samples.coop/Utilities/Net/SessionManager.cs
```

This class allows BossRoom to handle player session data, represented by a struct `T` implementing the `ISessionPlayerData` interface, by providing mechanisms to initialize, obtain and edit that data, and to associate it to a specific player. It also handles the clearing of data that is no longer used and the reinitialization of data between sessions.

> In this case, since game sessions are quite short, the session data is only cleared for disconnected players when a session ends, and when a new one begins. This makes sure that if a player disconnects during a session and then reconnects during the next session, the game properly treats it as a new connection. The definition of when a session ends and when a session starts might vary from game to game, but in BossRoom a session is considered to start after character selection and end when the players win or loose the game and enter the post-game scene. In other cases, one might want to add a timeout to session data and clear it after a specified time instead.

Reviewer: ```suggestion
In this case, since game sessions are quite short, the session data is only cleared for disconnected players when a session ends, and when a new one begins. This makes sure that if a player disconnects during a session and then reconnects during the next session, the game properly treats it as a new connection. The definition of when a session ends and when a session starts might vary from game to game, but in BossRoom a session is considered to start after character selection and end when the players win or loose the game and enter the post-game scene. In other cases, one might want to add a timeout to session data and clear it after a specified time instead.

This code is in Boss Room's utilities package so it can be easily reused.
```
